### ISSUE TITLE
RFC: DanceGRPO - Multi-Backend RL Training for Diffusion Models

### ISSUE THREAD

**Issue by Kirrito-k423** [verl-project/verl#5465]
## Summary

We propose contributing two production-ready recipes for training diffusion models with reinforcement learning using the **DanceGRPO** algorithm, targeting different backend infrastructures:

1. **MindSpeed-MM Backend** (PR #[50](https://github.com/verl-project/verl-recipe/pull/50)) - For Ascend NPU infrastructure
2. **FSDP Backend** (PR #[59](https://github.com/verl-project/verl-recipe/pull/59)) - For GPU infrastructure with pure PyTorch FSDP

### Related PRs
- #[50](https://github.com/verl-project/verl-recipe/pull/50): Recipe: Reinforcement learning for generative models using MindSpeed-mm as the backend (DanceGRPO)
- #[59](https://github.com/verl-project/verl-recipe/pull/59): Recipe: Reinforcement learning for generative models using FSDP as Backend (DanceGRPO)

## Motivation

Diffusion models have achieved remarkable success in image/video generation. However, aligning these models with human preferences through RLHF remains challenging due to:

- **High computational cost** of generating samples and computing gradients through the diffusion process
- **Lack of unified frameworks** that support multiple backends (NPU/GPU)
- **Complexity of distributed training** for large-scale diffusion models

Our work addresses these challenges by providing end-to-end recipes integrated with VeRL.

### 1. MMDanceGRPO (MindSpeed-MM Backend)

**PR**: #[50](https://github.com/verl-project/verl-recipe/pull/50)

**Key Features**:
- Integration with [MindSpeed-MM](https://gitcode.com/Ascend/MindSpeed-MM) for Ascend NPU training
- Wan2.2-5B as policy model with HPSv3 reward model
- FSDP/FSDP2 distributed training support
- Ulysses sequence parallelism for long sequences

**Architecture**:

```bash
RayDANCETrainer (Driver)
└── DiffusionActorRolloutWorker
├── HFRollout (Generation)
├── HPSv3 RM (Reward)
└── DataParallelPPOActor (GRPO Update)
└── MindSpeed-MM FSDP Backend
```

### 2. DanceGRPO-FSDP (FSDP Backend)
**PR**: #[59](https://github.com/verl-project/verl-recipe/pull/59)
**Key Features**:
- Pure PyTorch FSDP implementation, GPU-friendly
- mammothmoda as policy model with HPSv3 reward model
- Complete model implementation included (mammothmoda2)
- Supports both T2I and T2V training scenarios
## Algorithm Details
Both recipes implement **DanceGRPO** (Group Relative Policy Optimization for Diffusion):
1. **Rollout Generation**: Sample from diffusion model with log probability tracking
2. **Reward Computation**: HPSv3 scoring on generated samples
3. **GRPO Update**: 
   - Group-relative advantage normalization
   - PPO-style clipping on diffusion denoising ratios
   - Gradient accumulation and clipping
## Integration with VeRL
- Inherits from `RayPPOTrainer` for distributed orchestration
- Reuses VeRL's `DataProto` for data transfer
- Custom workers for diffusion-specific rollout and training
## References
- DanceGRPO: https://dancegrpo.github.io/
- MindSpeed-MM: https://gitcode.com/Ascend/MindSpeed-MM
- HPSv3 Reward Model: https://github.com/MizzenAI/HPSv3
- Wan2.2: https://github.com/Wan-Video/Wan2.2
- mammothmoda: https://github.com/bytedance/mammothmoda
## Questions for Community
1. Would the community be interested in unifying these two backends under a common abstraction?
2. Any preferences on naming conventions for the recipe directories?
3. Should we consider adding more reward model options beyond HPSv3?
---
We welcome feedback and suggestions from the community!

### Motivation

...

### Your contribution

...

**Comment by Kirrito-k423** [comment]
Excellent RFC! DanceGRPO for diffusion models is a valuable addition. A few comments:

**1. Backend unification**: Yes, a common abstraction would be beneficial. Consider creating a `DiffusionBackend` interface that both MindSpeed-MM and FSDP implementations conform to. This would allow users to switch backends without changing training code.

**2. Naming**: Suggest `recipes/dancegrpo_mindspeed` and `recipes/dancegrpo_fsdp` to clearly distinguish the backends.

**3. Reward models**: HPSv3 is a solid choice for image quality. For video, consider adding:
- CLIP-based temporal consistency rewards
- Optical flow-based motion quality rewards

**4. Future extensions**:
- T2V models would benefit from frame-level reward decomposition
- Consider adding classifier-free guidance (CFG) scale as a trainable parameter

Looking forward to both PRs being merged. This fills an important gap in veRL's model coverage.
